README
This is liwc, version 1.21.

liwc is a collection of programs to manipulate C source code,
originally written by Lars "liw" Wirzenius.

ccmtcnvt converts C++ comments (// foo) to C comments (/* foo */).
rmccmt removes C and C++ comments.
entrigraph converts single characters to trigraphs.
untrigraph converts trigraphs to single characters.
cstr prints string literals in C programs.

You need the Publib library (see http://code.google.com/p/publib/) and
a <unistd.h> that declares getopt. Once you have these (or hack the
source to avoid them), you can compile with "make" and install with
"make install".

Please file bug reports in the issue tracker at
  http://code.google.com/p/liwc/issues/list
or alternatively by electronic mail to
    antti-juhani@kaijanaho.fi

Development code is hosted in Git.  There's a code browser at
  http://code.google.com/p/liwc/source/browse/
and a local clone can be made by
  git clone https://code.google.com/p/liwc/

Releases will be posted at
  http://code.google.com/p/liwc/downloads/list


liwc is free software under the following copyright and license statement:

Copyright (c) 2012 Antti-Juhani Kaijanaho
Copyright (c) 1994-2003 Lars Wirzenius

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License, version 2, as
published by the Free Software Foundation.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License along
with this program; if not, write to the Free Software Foundation, Inc.,
51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.


ChangeLog
2003-02-06  Lars Wirzenius  <liw@iki.fi>

	* liwc.1: Wrote. Suggested by Martin Pool (mbp@sourcefrog.net).
	
	* Making release 1.20.

2003-01-19  Lars Wirzenius  <liw@iki.fi>

    	* Making release 1.9.

	* rmccmt.c, rmccmt.1: Added option -c to remove code, keep
	comments.

2001-09-23  Lars Wirzenius  <liw@iki.fi>

	* New version 1.8. The only difference is that the debian/* files
	have been removed, since I no longer maintain the Debian package.
	Antti-Juhani Kaijanaho has taken over the Debian package.


cstr.c
/*
 * Part of liwc, a collection of tools for manipulating C source code
 * Copyright (c) 1994-2003 Lars Wirzenius
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License,
 * version 2, as published by the Free Software Foundation.
 * 
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
 * 02110-1301 USA.
 */
/*
 * ccstr.c -- print out C string literals
 * Lars Wirzenius
 *
 * Usage:	cstr [-hv] [--help] [--version] [file...]
 * All output is to the standard output.
 */

#include <assert.h>
#include <limits.h>
#include <stdio.h>
#include <stdarg.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <publib.h>


/*
 * Special marks for input characters used in the description of the FSM.
 */
#define DFL	UCHAR_MAX+1	/* any character */
#define NONE	UCHAR_MAX+2	/* don't print any character */
#define SELF	UCHAR_MAX+3	/* print last input character */

/*
 * Possible states for the FSM.
 */
enum state { 
	code, chr_lit, chr_esc, str_lit, str_esc, slash, c_cmt, star
};

/*
 * Rules for describing the state changes and associated actions for the
 * FSM.
 */
struct rule {
	enum state state;
	int c;
	enum state new_state;
	int print1, print2, print3, print4;
};

/*
 * The FSM itself.
 */
static const struct rule fsm[] = {
	{ code,	   '"',  str_lit, NONE,	NONE, NONE, NONE	},
	{ code,    '\'', chr_lit, NONE,	NONE, NONE, NONE	},
	{ code,	   '/',  slash,   NONE,	NONE, NONE, NONE	},
	{ code,	   DFL,  code,    NONE,	NONE, NONE, NONE	},
			   
	{ str_lit, '\\', str_esc, SELF,	NONE, NONE, NONE	},
	{ str_lit, '"',  code,    '\n',	NONE, NONE, NONE	},
	{ str_lit, DFL,  str_lit, SELF,	NONE, NONE, NONE	},
			   
	{ str_esc, DFL,  str_lit, SELF,	NONE, NONE, NONE	},
			   
	{ chr_lit, '\\', chr_esc, NONE,	NONE, NONE, NONE	},
	{ chr_lit, '\'', code,    NONE,	NONE, NONE, NONE	},
	{ chr_lit, DFL,  chr_lit, NONE,	NONE, NONE, NONE	},
			   
	{ chr_esc, DFL,  chr_lit, NONE,	NONE, NONE, NONE	},
			   
	{ slash,   '/',  slash,   NONE,	NONE, NONE, NONE	},
	{ slash,   '*',  c_cmt,   NONE,	NONE, NONE, NONE	},
	{ slash,   DFL,  code,    NONE,	NONE, NONE, NONE	},

	{ c_cmt,   '*',  star,    NONE, NONE, NONE, NONE	},
	{ c_cmt,   DFL,  c_cmt,   NONE, NONE, NONE, NONE	},

	{ star,    '/',  code,    NONE, NONE, NONE, NONE	},
	{ star,    '*',  star,    NONE, NONE, NONE, NONE	},
	{ star,    DFL,  c_cmt,   NONE, NONE, NONE, NONE	},
};




static int cstr(FILE *f, char *filename, void *);

static const char usage[] = "[-hv] [file ...]\n";
static const char version[] = "version 1.0";


/*
 * Do the usual main() stuff: interpret options, process all input files.
 */
int main(int argc, char **argv) {
	int opt;

	set_progname(argv[0], "cstr");

	while ((opt = getopt(argc, argv, "hv")) != EOF) {
		switch (opt) {
		case 'h':
		case '?':
			errormsg(1, 0, "%s", usage);
		case 'v':
			errormsg(1, 0, "%s", version);
		}
	}
	
	if (main_filter(argc-optind, argv+optind, cstr, NULL) == -1)
		return EXIT_FAILURE;
	return EXIT_SUCCESS;
}


/*
 * Process one file.  Output is to stdout.  Return 0 for OK, -1 for error.
 */

#define print(x) (void)((x) != NONE && printf("%c", (x) == SELF ? c : (x)))

static int cstr(FILE *f, char *filename, void *dummy) {
	enum state state;
	const struct rule *p;
	int c, i;

	state = code;
	for (;;) {
		c = getc(f);
		if (c == EOF) {
			if (ferror(f)) {
				errormsg(0, -1, "error reading %s\n");
				return -1;
			}

			if (state != code) {
				errormsg(0, 0, "%s ends funnily\n", filename);
				return -1;
			}
			return 0; /* everything seems to be OK */
		}

		for (i = 0; i < sizeof(fsm) / sizeof(fsm[0]); ++i) {
			p = &fsm[i];
			if (p->state == state && (p->c == c || p->c == DFL)) {
				state = p->new_state;
				print(p->print1);
				print(p->print2);
				print(p->print3);
				print(p->print4);
				break;
			}
		}
	}
}


rmccmt.c
/*
 * Part of liwc, a collection of tools for manipulating C source code
 * Copyright (c) 2012 Antti-Juhani Kaijanaho
 * Copyright (c) 1994-2003 Lars Wirzenius
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License,
 * version 2, as published by the Free Software Foundation.
 * 
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
 * 02110-1301 USA.
 */
/*
 * Name:	rmccmt.c
 * Purpose:	Remove comments from C and C++ programs. Optionally removes
 *  	    	code and leaves the comment instead.
 * Author:	Lars Wirzenius
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <publib.h>

enum state { 
	code, 		/* In code, i.e., outside anything else */
	begcmt, 	/* '/' has been read, possible beginning of comment */
	incmt, 		/* Inside trad. C style comment */
	endcmt, 	/* '*' has been read, possible end of comment */
	incppcmt, 	/* Inside C++ style comment */
	str, 		/* Inside string literal */
	strquote, 	/* Backslash-escaped character inside str. literal */
	chr, 		/* Inside character literal */
	chrquote 	/* Backslash-escaped character inside char. literal */
};

static int keepnl;
static int rmcode;

static void print_cmt(int c)
{
    if (rmcode)
    	putchar(c);
}

static void print_code(int c)
{
    if (!rmcode)
    	putchar(c);
}

static int rmcmt(FILE *f, char *filename, void *dummy) {
	enum state st;
	int c;

	st = code;
	while ((c = getc(f)) != EOF) {
		switch (st) {
		case code:
			switch (c) {
			case '/':
				st = begcmt;
				break;
			case '"':
				st = str;
				print_code(c);
				break;
			case '\'':
				st = chr;
				print_code(c);
				break;
			default:
				print_code(c);
				break;
			}
			break;

		case begcmt:
			switch (c) {
			case '/':
				st = incppcmt;
				print_cmt('/');
				print_cmt('/');
				break;
			case '*':
				st = incmt;
				print_cmt('/');
				print_cmt('*');
				break;
			case '"':
				st = str;
				print_code('/');
				print_code(c);
				break;
			case '\'':
				st = chr;
				print_code('/');
				print_code(c);
				break;
			default:
				print_code('/');
				st = code;
				print_code(c);
				break;
			}
			break;

		case incmt:
			print_cmt(c);
			if (c == '\n' && keepnl && !rmcode)
				putchar('\n');
			if (c == '*')
				st = endcmt;
			break;

		case endcmt:
			print_cmt(c);
			if (c == '\n' && keepnl && !rmcode)
				putchar('\n');
			if (c == '/') {
				st = code;
				if (rmcode)
				    putchar('\n'); /* next cmt on new line */
				else
				    putchar(' '); /* replace cmt with space */
			} else if (c != '*')
				st = incmt;
			break;

		case incppcmt:
		    	print_cmt(c);
			if (c == '\n') {
				st = code;
				if (!rmcode)
				    putchar('\n');
			}
			break;

		case str:
			print_code(c);
			if (c == '\\')
				st = strquote;
			else if (c == '"')
				st = code;
			break;

		case strquote:
			print_code(c);
			st = str;
			break;

		case chr:
			print_code(c);
			if (c == '\\')
				st = chrquote;
			else if (c == '\'')
				st = code;
			break;

		case chrquote:
			print_code(c);
			st = chr;
			break;
		}
	}

	if (st != code) {
		errormsg(0, 0, 
			"input is malformed, ends inside comment or literal");
		return -1;
	}
	
	if (fflush(stdout) == EOF || ferror(stdout))
		return -1;
	return 0;
}


int main(int argc, char **argv) {
	set_progname(argv[0], "rmccmt");
	
	while (argc > 1 && argv[1][0] == '-') {
	    	if (strcmp(argv[1], "-n") == 0)
		    keepnl = 1;
		else if (strcmp(argv[1], "-c") == 0)
		    rmcode = 1;
		--argc;
		++argv;
	}
	if (main_filter(argc-1, argv+1, rmcmt, NULL) == -1)
		return EXIT_FAILURE;
	return EXIT_SUCCESS;
}


entrigraph.c
/*
 * Part of liwc, a collection of tools for manipulating C source code
 * Copyright (c) 1994-2003 Lars Wirzenius
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License,
 * version 2, as published by the Free Software Foundation.
 * 
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
 * 02110-1301 USA.
 */
/*
 * Name:	entrigraph.c
 * Purpose:	Put trigraphs into C source code file.
 * Author:	Lars Wirzenius
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <publib.h>

int entrigraph(FILE *f, char *fname, void *dummy) {
	int c;
	char tri[]       = "()<>!'-=/";  /* list of trigraphs... */
	char realchars[] = "[]{}|^~#\\"; /* ...and corresponding chars */
	char *p;

	while ((c = getc(f)) != EOF) {
		if ((p = strchr(realchars, c)) == NULL)
			putchar(c);
		else
			printf("?\?%c", tri[p-realchars]);
	}
	return 0;
}

int main(int argc, char **argv) {
	set_progname(argv[0], "entrigraph");
	if (main_filter(argc-1, argv+1, entrigraph, NULL) == -1)
		exit(EXIT_FAILURE);
	exit(0);
}


untrigraph.c
/*
 * Part of liwc, a collection of tools for manipulating C source code
 * Copyright (c) 1994-2003 Lars Wirzenius
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License,
 * version 2, as published by the Free Software Foundation.
 * 
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
 * 02110-1301 USA.
 */
/*
 * Name:	untrigraph.c
 * Purpose:	Remove trigraphs from C code.
 * Author:	Lars Wirzenius
 */

#include <assert.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <publib.h>

int untrigraph(FILE *, char *, void *);

int main(int argc, char **argv) {
	set_progname(argv[0], "untrigraph");
	if (main_filter(argc-1, argv+1, untrigraph, NULL) == -1)
		return EXIT_FAILURE;
	return EXIT_SUCCESS;
}


/*
 * Name:	untrigraph
 * Purpose:	Remove trigraphs from one file, output to stdout.
 * Description:	A sequence of four characters (c0, c1, c2, c3) is a
 *		trigraph, if c0 != '\\', c1==c2=='?' and c3 is one of the
 *		characters inside the string "()<>!'-=/".
 *
 *		In the code below, c0 is the previous character we printed
 *		(possibly the result of a trigraph conversion).  
 */
int untrigraph(FILE *f, char *fname, void *dummy) {
	int c0, c1, c2, c3;
	char tri[]       = "()<>!'-=/";  /* list of trigraphs... */
	char realchars[] = "[]{}|^~#\\"; /* ...and corresponding chars */
	char *p;

	c0 = c2 = c3 = EOF;
	p = NULL;

	if ((c1=getc(f)) != EOF && (c2=getc(f)) != EOF)
		c3 = getc(f);
	while (c1 != EOF) {
		if (c0 == '\\' || c1 != '?' || c2 != '?' ||
		    (c3 != EOF && (p = strchr(tri, c3)) == NULL)) {
			putchar(c1);
			c0 = c1;
			c1 = c2;
			c2 = c3;
			if (c3 != EOF)
				c3 = getc(f);
		} else {	/* it's a trigraph! */
			assert(p != NULL);
			assert(p >= tri && p < tri + sizeof(tri) - 1);
			c0 = realchars[p-tri];
			putchar(c0);
			c1 = c2 = c3 = EOF;
			if ((c1=getc(f)) != EOF && (c2=getc(f)) != EOF)
				c3 = getc(f);
		}
	}
	if (ferror(f))
		return -1;
	if (fflush(stdout) == EOF || ferror(stdout))
		return -1;
	return 0;
}


ccmtcnvt.c
/*
 * Part of liwc, a collection of tools for manipulating C source code
 * Copyright (c) 2012 Antti-Juhani Kaijanaho
 * Copyright (c) 1994-2003 Lars Wirzenius
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License,
 * version 2, as published by the Free Software Foundation.
 * 
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
 * 02110-1301 USA.
 */
/*
 * ccmtcnvt.c -- convert C++ comments to C comments
 * Lars Wirzenius
 *
 * Usage:	ccmtcnvt [-hv] [--help] [--version] [file...]
 * All output is to the standard output.
 *
 * This  program  converts the C++ style comments into traditional C style
 * comments.   C++  allows  comments  of  that begin  with  `//'  and end
 * with the next newline, but they weren't introduced to the C language
 * until the  1999  version  of the C standard.  Some C compilers alre3ady
 * accept these, but for those who don't, you can use  this  utility to
 * convert them.
 * 
 * This is written as a finite state machine (FSM), which reads (presumably
 * correct) C code, and when it finds a // that is not inside a comment, or
 * string or character literal, it converts it to a C start of comment and
 * adds a space and a comment delimiter (unreproducible here :-) before the
 * next newline.
 *
 * The command line options require POSIX 1003.2 to work, because they
 * assume that getopt is declared by <unistd.h>.
 *
 * Known problems: 
 *
 * a) Include file names are not recognized being special, so that
 *	
 *	#include <sys//stat.h>
 *
 * will be mangled.  I've never seen such names used in source code, though,
 * so I don't think ignoring them will cause too much harm.
 *
 * b) Does not understand trigraphs or digraphs.
 *
 * c) Does not understand line continuation (\ at end of line) outside
 * string literals and character constants.
 *
 * Patches for fixing these are welcome. They're not common enough for me
 * to worry about them, at least for now.
 */

/*
 * Neat trick dept.: the following expression will evaluate to 0 if
 * C++ comments are not supported, and 1 if they are.  (x is here *;
 * can't use * because it would terminate the comment.)
 *
 *	int cppcmts = 1//xx/2
 *		;
 *
 * or use it like this:
 *
 *	#if 1//xx/2
 *	#error This C compiler is confused about comment syntax
 *	#endif
 */

#include <assert.h>
#include <errno.h>
#include <limits.h>
#include <stdio.h>
#include <stdarg.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <publib.h>


/*
 * Special marks for input characters used in the description of the FSM.
 */
#define DFL	(int)UCHAR_MAX+1	/* any character */
#define NONE	(int)UCHAR_MAX+2	/* don't print any character */
#define SELF	(int)UCHAR_MAX+3	/* print last input character */

/*
 * Possible states for the FSM.
 */
enum state { 
	code, chr_lit, chr_esc, str_lit, str_esc, 
	slash, cpp_cmt, cpp_ast, c_cmt, star
};

/*
 * Rules for describing the state changes and associated actions for the
 * FSM.
 */
struct rule {
	enum state state;
	int c;
	enum state new_state;
	int print1, print2, print3, print4;
	const char *warning;
};

/*
 * The FSM itself.
 */
static const struct rule fsm[] = {
	{ code,	   '"',  str_lit, SELF,	NONE, NONE, NONE, NULL	},
	{ code,    '\'', chr_lit, SELF,	NONE, NONE, NONE, NULL	},
	{ code,	   '/',  slash,   NONE,	NONE, NONE, NONE, NULL	},
	{ code,	   DFL,  code,    SELF,	NONE, NONE, NONE, NULL	},
			   
	{ str_lit, '\\', str_esc, SELF,	NONE, NONE, NONE, NULL	},
	{ str_lit, '"',  code,    SELF,	NONE, NONE, NONE, NULL	},
	{ str_lit, DFL,  str_lit, SELF,	NONE, NONE, NONE, NULL	},
			   
	{ str_esc, DFL,  str_lit, SELF,	NONE, NONE, NONE, NULL	},
			   
	{ chr_lit, '\\', chr_esc, SELF,	NONE, NONE, NONE, NULL	},
	{ chr_lit, '\'', code,	  SELF,	NONE, NONE, NONE, NULL	},
	{ chr_lit, DFL,	 chr_lit, SELF,	NONE, NONE, NONE, NULL	},
			   
	{ chr_esc, DFL,	 chr_lit, SELF,	NONE, NONE, NONE, NULL	},
			   
	{ slash,   '/',	 cpp_cmt, '/',	'*',  NONE, NONE, NULL	},
	{ slash,   '*',	 c_cmt,	  '/',	'*',  NONE, NONE, NULL	},
	{ slash,   DFL,	 code,	  '/',	SELF, NONE, NONE, NULL	},

	{ cpp_cmt, '\n', code,	  ' ',	'*',  '/',  SELF, NULL	},
	{ cpp_cmt, '*',	 cpp_ast, SELF,	NONE, NONE, NONE, NULL	},
	{ cpp_cmt, DFL,	 cpp_cmt, SELF, NONE, NONE, NONE, NULL	},

	{ cpp_ast, '\n', code,	  ' ',	'*',  '/',  SELF, NULL	},
	{ cpp_ast, '*',	 cpp_ast, SELF,	NONE, NONE, NONE, NULL	},
	{ cpp_ast, '/',	 cpp_cmt, ' ',	SELF, NONE, NONE,
	  "converted a '*/' inside a C++ comment to '* /'."	},
	{ cpp_ast, DFL,	 cpp_cmt, SELF, NONE, NONE, NONE, NULL	},

	{ c_cmt,   '*',	 star,	  SELF, NONE, NONE, NONE, NULL	},
	{ c_cmt,   DFL,	 c_cmt,	  SELF, NONE, NONE, NONE, NULL	},

	{ star,	   '/',	 code,	  SELF, NONE, NONE, NONE, NULL	},
	{ star,	   '*',	 star,	  SELF, NONE, NONE, NONE, NULL	},
	{ star,	   DFL,	 c_cmt,	  SELF, NONE, NONE, NONE, NULL	},
};


static int convert(FILE *f, char *filename, void *);

static char usage[] = "Usage: %s [-hv] [file ...]\n";
static char version[] = "version 1.1\n";


/*
 * Do the usual main() stuff: interpret options, process all input files.
 */
int main(int argc, char **argv) {
	int opt;

	set_progname(argv[0], "ccmtcnvt");

	while ((opt = getopt(argc, argv, "hv")) != EOF) {
		switch (opt) {
		case 'h':
		case '?':
			errormsg(1, 0, usage, get_progname());
		case 'v':
			errormsg(1, 0, "%s", version);
		}
	}

	if (main_filter(argc-optind, argv+optind, convert, NULL) == -1)
		return EXIT_FAILURE;
	return EXIT_SUCCESS;
}

#define print(x) (void)((x) != NONE && printf("%c", (x) == SELF ? c : (x)))
static int convert(FILE *f, char *filename, void *dummy) {
	enum state state;
	const struct rule *p;
	int c, i;

	state = code;
	for (;;) {
		c = getc(f);
		if (c == EOF) {
			if (ferror(f)) {
				errormsg(0, errno, "error reading %s\n");
				return -1;
			}

			switch (state) {
			case cpp_cmt:
				/* need to terminate the last // comment */
				printf(" */");
				break;
			case code:
				/* do nothing 'cause all's good */
				break;
			default:
				errormsg(0, 0, "%s ends funnily\n", filename);
				return -1;
			}
	
			if (fflush(stdout) == EOF || ferror(stdout))
				return -1;
			return 0; /* everything seems to be OK */
		}

		for (i = 0; i < sizeof(fsm) / sizeof(fsm[0]); ++i) {
			p = &fsm[i];
			if (p->state == state && (p->c == c || p->c == DFL)) {
				state = p->new_state;
				print(p->print1);
				print(p->print2);
				print(p->print3);
				print(p->print4);
				if (p->warning != NULL)
					errormsg(0, 0, "%s", p->warning);
				break;
			}
		}
	}
	/*NOTREACHED*/
}
